### PR TITLE
Updating authorization building block with generic UserHasPermission attribute

### PR DESCRIPTION
**Pull request by vsljivic**

### CODE REVIEW

File: Enigmatry.Entry.AspNetCore.Authorization/Requirements/AuthenticatedUserRequirementHandler.cs [hunk 1]
```diff
@@ -33,7 +34,7 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
             var requirementName = nameof(requirement);
```
Comment (vsljivic) [inline Enigmatry.Entry.AspNetCore.Authorization/Requirements/AuthenticatedUserRequirementHandler.cs:34]: nameof(requirement) is just "requirement"
implement .ToString() on the requirement so more data from the requirement can be logged
add userid to the logs.
"user with id does not have correct rights: '...' for access this controller: '...' "

File: Enigmatry.Entry.AspNetCore.Authorization/AuthorizationStartupExtensions.cs [hunk 2]
```diff
@@ -14,14 +15,14 @@ public static class AuthorizationStartupExtensions
         bool enable) =>
         !enable ? builder : builder.RequireAuthorization();
 
-    public static AuthorizationBuilder AppAddAuthorization(this IServiceCollection services, bool enable)
+    public static AuthorizationBuilder AppAddAuthorization<T>(this IServiceCollection services, bool enable = true)
```
Comment (vsljivic) [inline Enigmatry.Entry.AspNetCore.Authorization/AuthorizationStartupExtensions.cs:18]: Add xml docs.
Remove 'enable' flag.

File: Enigmatry.Entry.AspNetCore.Authorization/UserHasPermissionPolicyProvider.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            var requirement =
+                new UserHasPermissionRequirement<T>(PermissionsConverter<T>.ParseFromPolicyName(UserHasPermissionAttribute<T>.PolicyPrefix, policyName));
+            return Task.FromResult(new AuthorizationPolicyBuilder().AddRequirements(requirement).Build())!;
+        }
+
+        return GetDefaultPolicyAsync()!;
+    }
+
+    // TODO:This also adds the policy to swagger and healthcheck endpoints. Need to find a solution
```
Comment (vsljivic) [inline Enigmatry.Entry.AspNetCore.Authorization/UserHasPermissionPolicyProvider.cs:21]: Test this TODO via the SampleApp.

File: Enigmatry.Entry.AspNetCore.Authorization/UserHasPermissionPolicyProvider.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    public Task<AuthorizationPolicy> GetDefaultPolicyAsync() => Task.FromResult(new AuthorizationPolicyBuilder().RequireAuthenticatedUser().Build());
+
+    public Task<AuthorizationPolicy?> GetFallbackPolicyAsync() => GetDefaultPolicyAsync()!;
```
Comment (vsljivic) [inline Enigmatry.Entry.AspNetCore.Authorization/UserHasPermissionPolicyProvider.cs:24]: Should fallback be implemented? Is this the right way to force authentication for all endpoints?

File: Enigmatry.Entry.AspNetCore.Authorization/Attributes/PolicyNames.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Enigmatry.Entry.AspNetCore.Authorization.Attributes;
+
+internal static class PolicyNames
+{
+    private const char PermissionsDelimiter = ',';
+
+    public static string Format(string policyPrefix, IEnumerable<string> permissions) =>
+        $"{policyPrefix}{string.Join(PermissionsDelimiter, permissions)}";
```
[the remark this answers is not in the document]
Comment (vsljivic) [reply Enigmatry.Entry.AspNetCore.Authorization/Attributes/PolicyNames.cs]: This class is used internally in the authorization building block to encode permissions to policy name, and later from policy name back to list of permissions. It does not change order or anything. List specified in the UserHasPermission attribute is passed in the same order to a Project's implementation of the IAuthorizationProvider.

**Review by avanwaas — APPROVED** [review]

**Review by zsrdjan — APPROVED** [review]

**Pull request merged**